ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -663,35 +663,46 @@ class JITModBuilder {
     Function *F = AOTMod->getFunction(YK_NEW_CONTROL_POINT);
     User *CallSite = F->user_back();
     CallInst *CPCI = cast<CallInst>(CallSite);
-    Type *OutputStructTy = CPCI->getType();
-
-    // When assembling a trace, we start collecting instructions below the
-    // control point and finish above it. This means that alloca'd variables
-    // become undefined (as they are defined outside of the trace) and thus
-    // need to be remapped to the input of the compiled trace. SSA values
-    // (from the same block as the control point) remain correct as phi nodes
-    // at the beginning of the trace automatically select the appropriate input
-    // value.
+    Type *YkCtrlPointVarsPtrTy = F->getArg(1)->getType();
+    assert(YkCtrlPointVarsPtrTy->isPointerTy());
+
+    // When executing the interpreter loop AOT code, execution starts before
+    // the control point and flows below it, but when assembling a trace
+    // control flow is upside-down: we start collecting instructions below the
+    // control point and finish above it.
+    //
+    // This has implications: without extra logic, alloca'd variables become
+    // undefined (as they are defined outside of the trace) and thus need to be
+    // remapped to the input of the compiled trace. SSA values (from the same
+    // block as the control point) remain correct as phi nodes at the beginning
```

Comment:
Since everything is a pointer into a struct now, can there even be phi nodes now?

Reply:
> Since everything is a pointer into a struct now, can there even be phi nodes now?

Not everything is a pointer. For example, an `i32` is copied into and out of the struct by value. It's the struct that is passed by pointer.

Also, the data in the struct is only used by the insides of the control point. The rest of the time, the "raw" values are used, so yes, we do still see PHI nodes for those.

e.g. in the `simple_interp_loop1` test:

```
25:
  %26 = phi i32 [ %19, %13 ], [ 0, %100 ], [ %103, %101 ], [ %96, %92 ], !dbg !82
  ...
  %43 = getelementptr %YkCtrlPointVars, %YkCtrlPointVars* %3, i32 0, i32 14, !dbg !93
  store i32 %26, i32* %43, align 4, !dbg !93
  ...
  call void @yk_new_control_point(%struct.YkLocation* %28, %YkCtrlPointVars* %3), !dbg !93
  ...
  %100 = getelementptr %YkCtrlPointVars, %YkCtrlPointVars* %0, i32 0, i32 14, !dbg !8
  %101 = load i32, i32* %100, align 4, !dbg !8
  ...
```

Hopefully that makes sense?

Reply:
Yes, that makes sense.